RIPS/rip-9999.md

```diff
@@ -0,0 +1,307 @@
+
+This proposal is intended to prevent the fragmentation and inconsistencies that may arise as a result of implementing
+multiple independent standalone proposals by combining all of them into a unified design.
+
+Here is the list of EIPs that propose adding features to EOA accounts that are addressed by this document:
+
+* [EIP-2711: Sponsored, expiring and batch transactions](https://eips.ethereum.org/EIPS/eip-2711)
+* [EIP-3074: AUTH and AUTHCALL opcodes](https://eips.ethereum.org/EIPS/eip-3074)
+* [EIP-5003: Insert Code into EOAs with AUTHUSURP](https://eips.ethereum.org/EIPS/eip-5003)
+* [EIP-5081: Expirable Transaction](https://eips.ethereum.org/EIPS/eip-5081)
+* [EIP-5806: Delegate transaction](https://eips.ethereum.org/EIPS/eip-5806)
+* [EIP-7377: Migration Transaction](https://eips.ethereum.org/EIPS/eip-7377)
+* [EIP-7553: Separated Payer Transaction](https://eips.ethereum.org/EIPS/eip-7553)
+
+Each one of these proposals introduces features that can benefit both existing users and potential future users of
+the Ethereum blockchain in multiple ways.
+
+However, together these proposals consist of 3 different transaction types and 3 opcodes, which is probably unsustainable. The transaction types are incompatible with each other, forcing the user to only use one.
+
+None of these proposals supports Gas Abstraction mechanism that would be conceptually similar to "Paymasters" - smart
+contracts capable of paying transaction gas, based on some on-chain validity function, but do not gain any
+privileged access to sender's funds.
+This is very disappointing given that the concept of "Paymasters" has being widely used in production by
+[ERC-4337: Account Abstraction Using Alt Mempool](https://eips.ethereum.org/EIPS/eip-4337)
+and even
+[ERC-1613: Gas Stations Network](https://eips.ethereum.org/EIPS/eip-1613)
+before that.
+
+None of these proposals fully support validity time ranges, which enable both scheduled and expiring transactions,
+adding a noticeable improvement to end-user experience with a relatively small amount of effort.
+
+And most importantly, none of these proposals form a part of a bigger vision.
+Instead, these proposals are isolated patches that can be applied in a single design iteration.
+These proposals cannibalize each other's use-cases, but do not fully resolve the pain points Ethereum users are
+currently suffering from, and they do not fully align with any kind of vision for a Full Native Account Abstraction.
+
+## Specification
+The aforementioned EIPs can be implemented by the following transaction subtypes:
+
+### Shared RIP-9999 transaction payload
+
+The required synergy of the proposed transaction subtypes is achieved by defining a shared base structure of
+a transaction.
+This will ensure the only differences between these subtypes are the ones strictly required
+by their semantics instead of being caused by different tastes or variations in implementation details.
```

Comment:
Why create a new tx type within a type? Clients will need to now parse out the 2718 tx type and then do a match on the number of RLP elements in the list to figure out what subtype it is. This implicitly means that subtypes cannot have the same length of parameters, otherwise it will not be possible to determine which tx type to decode to.

Reply:
This RIP does propose a set of potential transaction types but we wanted to avoid hogging the entire types-space with our proposals, and these types share a lot in common.
So the idea behind defining a "subtype" if basically to extend the EIP-2718 type field to be 16 bit. This allows us to create 256 "subtypes" within the "Type 4" transaction without a risk interfering with any other EIP or RIP. Clients will need to parse out the EIP-2718 transaction type and, if it equals 4, also parse out the AA transaction subtype.

Something similar will probably need to happen in Solidity, where specific structs will need to be strictly extending the base transaction struct. This part is still missing from the document.

Does this approach make sense to you? Do you have a suggestion on how we can address this issue better?
Thanks!

Reply:
Since there are still ~120 available EIP-2718 tx types I think you should use them directly. The subtype adds a lot of complexity to clients and tooling. If we ever get into a situation where we're nearing the end of the available range, we can institute subtype format.